moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp

```diff
@@ -370,35 +373,96 @@ class Ros2ControlManager : public moveit_controller_manager::MoveItControllerMan
     return c;
   }
 
+  static void simplifyControllerActivationDeactivation(std::vector<std::string>& activate_controllers,
+                                                       std::vector<std::string>& deactivate_controllers)
+  {
+    // Activation/deactivation is expected to be disjoint. For example, if controller B is a dependency of A (A chains
+    // to B) but controller B is also a dependency of C (B chains to B), then the switch from A->B to C->B would cause B
+    // to be in both the activation and deactivate list. This causes ROS2 control to through an error and reject the
+    // switch. The simplifyControllerActivationDeactivation function adds the logic needed to avoid this from happening.
+
+    // Convert vectors to sets for uniqueness
+    std::unordered_set set1(activate_controllers.begin(), activate_controllers.end());
+    std::unordered_set set2(deactivate_controllers.begin(), deactivate_controllers.end());
+
+    // Find common elements
+    std::unordered_set<std::string> common;
+    for (const auto& str : set1)
+    {
+      if (set2.count(str))
+      {
+        common.insert(str);
+      }
+    }
+
+    // Remove common elements from both sets
+    for (const auto& str : common)
+    {
+      set1.erase(str);
+      set2.erase(str);
+    }
+
+    // Convert sets back to vectors
+    activate_controllers.assign(set1.begin(), set1.end());
+    deactivate_controllers.assign(set2.begin(), set2.end());
+  }
+
   /**
    * \brief Filter lists for managed controller and computes switching set.
    * Stopped list might be extended by unsupported controllers that claim needed resources
-   * @param activate vector of controllers to be activated
-   * @param deactivate vector of controllers to be deactivated
+   * @param activate_base vector of controllers to be activated
+   * @param deactivate_base vector of controllers to be deactivated
    * @return true if switching succeeded
    */
   bool switchControllers(const std::vector<std::string>& activate_base,
                          const std::vector<std::string>& deactivate_base) override
   {
-    // add controller dependencies
-    std::vector<std::string> activate = activate_base;
-    std::vector<std::string> deactivate = deactivate_base;
-    for (auto controllers : { &activate, &deactivate })
-    {
-      auto queue = *controllers;
+    // add_all_dependencies traverses the provided dependency map and appends the results to the controllers vector.
+    auto add_all_dependencies = [](const std::unordered_map<std::string, std::vector<std::string>>& dependencies,
```

Comment:
Curious as to why this is a lambda and not a function?

Reply:
I am using a lambda because I want to reuse this code but only in the scope of this method.

Reply:
I apparently cannot resolve my own comments

Reply:
@mikeferguson with the lightning reflexes